fix: ensure correct MSSQL parameter conversion in where conditions

sudhirt4 opened the pull request:
Fixes input parameter conversion in SelectQueryBuilder when using an MSSQL connection.

Closes #11285

### Description of change

Instead of passing raw input paramters as it is this change uses the available driver.parametrizeValue method to convert parameters into MSSQL format. 

This implementation was referred from existing usage in [UpdateQueryBuilder](https://github.com/typeorm/typeorm/blob/master/src/query-builder/UpdateQueryBuilder.ts#L553) and [InsertQueryBuilder](https://github.com/typeorm/typeorm/blob/master/src/query-builder/InsertQueryBuilder.ts#L887)

Please note that this issue currently only fixes where condition when used under `.find` method. On testing it was not working for `.createQueryBuilder(..).where()` usage. Could use some insights on a proper approach to implement this parameter transformation.

### Pull-Request Checklist
- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #11285`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change (N/A)
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

sgarner commented:
So this is tangentially related to https://github.com/typeorm/typeorm/pull/10161

For clarity, the problem is that MSSQL requires query parameters to be typed, and if no type is given then the underlying node-mssql driver defaults to NVARCHAR. If the column is actually a VARCHAR and is indexed, SQL Server won't use the index due to the type mismatch.

sgarner commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -4278,6 +4280,12 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                                       column.transformer,
                                       parameterValue,
                                   ))
+                    } else {
```
Are you sure this path should be in the else branch here?

I think the value needs to be 'parametrized' [sic] regardless of whether the column has a value transformer or not. The transformer should be applied first, and then parameterization.

sudhirt4 replied:
Makes sense, updated with changes✅ 

I was worried if there would be some return type from transformer that parametrize function could not handle but seems like it's working just fine.

sgarner reviewed (changes requested):
@sudhirt4 I've run the tests, and some are failing - can you have a look at it?

sudhirt4 commented:
> @sudhirt4 I've run the tests, and some are failing - can you have a look at it?

Updated with the changes. 

Figured out that params of type FindOperator had to be [handled separately](https://github.com/typeorm/typeorm/pull/11298/files#diff-52bea788267888fdbb5f795f0dfb61504f16bd43e0afdfa28848abb4435121b5R4290). Instead of changing other functions to handle this I have adjusted it here following the approach used in transformer just above this. Let me know if this pattern is acceptable.

Have updated the tests to include these scenarios as well

sgarner commented on src/query-builder/SelectQueryBuilder.ts:
```diff
@@ -4280,6 +4282,24 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
                                   ))
                     }
 
+                    if (
+                        this.connection.driver.options.type === "mssql" &&
+                        parameterValue.type !== "raw"
+                    ) {
+                        const driver = this.connection.driver as SqlServerDriver
+                        if (parameterValue instanceof FindOperator) {
+                            parameterValue.transformValue({
+                                to: (v) => driver.parametrizeValue(column, v),
+                                from: (v) => v,
+                            })
+                        } else {
+                            parameterValue = driver.parametrizeValue(
+                                column,
+                                parameterValue,
+                            )
+                        }
+                    }
```
Although this works, the comparison against `parameterValue.type` is premature; `parameterValue` might be anything but the `type` property comparison only applies when it is a `FindOperator` instance. Suggest the raw check be moved to the inner condition.

Also adding a comment to explain what this block does:

```suggestion
                    // MSSQL requires parameters to carry extra type information
                    if (this.connection.driver.options.type === "mssql") {
                        const driver = this.connection.driver as SqlServerDriver
                        if (parameterValue instanceof FindOperator) {
                            if (parameterValue.type !== 'raw') {
                                parameterValue.transformValue({
                                    to: (v) => driver.parametrizeValue(column, v),
                                    from: (v) => v,
                                })
                            }
                        } else {
                            parameterValue = driver.parametrizeValue(
                                column,
                                parameterValue,
                            )
                        }
                    }
```

sudhirt4 replied:
Updated ✅ I amended by commit to include these changes

sgarner reviewed (approved):
Thank you @sudhirt4 for your effort on this!

sgarner reviewed (approved):
LGTM :tada:
